src/components/tl/mlx5/mcast/tl_mlx5_mcast_progress.c

```diff
@@ -0,0 +1,386 @@
+/**
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast_progress.h"
+
+static ucc_status_t ucc_tl_mlx5_mcast_recv_completion(ucc_tl_mlx5_mcast_p2p_completion_obj_t *obj);
+
+static ucc_status_t ucc_tl_mlx5_mcast_send_completion(ucc_tl_mlx5_mcast_p2p_completion_obj_t *obj);
+
+static ucc_status_t ucc_tl_mlx5_mcast_dummy_completion(ucc_tl_mlx5_mcast_p2p_completion_obj_t *obj)
+{
+    return UCC_OK;
+}
+
+static ucc_tl_mlx5_mcast_p2p_completion_obj_t dummy_completion_obj = {
+    .compl_cb = ucc_tl_mlx5_mcast_dummy_completion,
+};
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_resend_packet_reliable(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                                    int p2p_pkt_id)
+{
+    uint32_t          psn = comm->p2p_pkt[p2p_pkt_id].psn;
+    struct pp_packet *pp  = comm->r_window[psn % comm->wsize];
+    ucc_status_t      status;
+
+    assert(pp->psn == psn);
+    
+    tl_trace(comm->lib, "[comm %d, rank %d] Send data NACK: to %d, psn %d, context %ld\n",
+                         comm->comm_id, comm->rank,
+                         comm->p2p_pkt[p2p_pkt_id].from, psn, pp->context);
+
+    status = comm->params.p2p_iface.send_nb((void*) (pp->context ? pp->context : pp->buf),
+                                            pp->length, comm->p2p_pkt[p2p_pkt_id].from,
+                                            comm->p2p_ctx, &dummy_completion_obj);
```

Comment:
Why is it necessary here and not for recv_nb?

Reply:
This depends on what we wanna do when the send is done. We use a dummy object here because once the entire packet of actual user data is sent out, then there is nothing to do for that packet on the sender side, hence we use a dummy completion object on the sender side. 
For recv_nb, we need to have a mechanism to handle the received packet once it arrives. Therefore, we do not use a dummy object, and we use a real completion method to handle the received packet.